File: tests/prolog/event-calculus.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { Prolog } from 'trealla';
import type { Answer } from 'trealla';
import { readFileSync } from 'fs';
import { join } from 'path';

/**
 * Helper: collect all answers from a query generator into an array.
 */
async function collectAnswers(gen: AsyncGenerator<Answer>): Promise<Answer[]> {
  const results: Answer[] = [];
  for await (const answer of gen) {
    results.push(answer);
  }
  return results;
}

describe('Event Calculus Prolog', () => {
  let prolog: Prolog | null;

  beforeEach(async () => {
    try {
      prolog = new Prolog({
        library: '/library'
      });
      await prolog.init();

      // Load Event Calculus rules if available
      try {
        const ecRules = readFileSync(join(process.cwd(), 'src/persistence/prolog-native.ts'), 'utf8');
        const prologCode = ecRules.match(/const PROLOG_EC_RULES = `([^`]+)`/)?.[1];

        if (prologCode) {
          await prolog.consultText(prologCode);
        }
      } catch {
        // EC rules file not found or no embedded Prolog code, that's fine
      }
    } catch {
      // If Prolog initialization fails, skip tests
      prolog = null;
    }
  });

  describe('Basic Event Calculus', () => {
    it('should handle happens/2 facts', async () => {
      if (!prolog) {
        console.log('Skipping Prolog test - initialization failed');
        return;
      }

      await prolog.consultText(`
        happens(event(alice, became, ceo), '2024-01-01').
        happens(event(bob, joined, company), '2024-02-01').
      `);

      const results = await collectAnswers(prolog.query('happens(event(alice, became, ceo), Date).'));
      expect(results).toBeDefined();
      expect(results.length).toBeGreaterThan(0);
      if (results[0].status === 'success') {
        expect(results[0].answer.Date).toBeDefined();
      }
    });

    it('should handle initiates/3 rules', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        initiates(event(Person, became, Role), role(Person, Role), _).
        initiates(event(Person, learned, Skill), knows(Person, Skill), _).
      `);

      const results = await collectAnswers(prolog.query('initiates(event(alice, became, ceo), Fluent, _).'));
      expect(results).toBeDefined();
      if (results.length > 0 && results[0].status === 'success') {
        expect(results[0].answer.Fluent).toBeDefined();
      }
    });

    it('should handle terminates/3 rules', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        terminates(event(Person, quit, Role), role(Person, Role), _).
        terminates(event(Person, moved_from, Location), lives_in(Person, Location), _).
      `);

      const results = await collectAnswers(prolog.query('terminates(event(alice, quit, manager), Fluent, _).'));
      expect(results).toBeDefined();
    });
  });

  describe('Fluent tracking', () => {
    it('should track holds_at/2', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        happens(event(alice, became, ceo), '2024-01-01').
        initiates(event(Person, became, Role), role(Person, Role), _).

        holds_at(Fluent, Date) :-
          happens(Event, StartDate),
          initiates(Event, Fluent, _),
          StartDate @=< Date,
          \\+ (
            happens(TermEvent, TermDate),
            terminates(TermEvent, Fluent, _),
            StartDate @< TermDate,
            TermDate @=< Date
          ).
      `);

      const result = await prolog.queryOnce('holds_at(role(alice, ceo), "2024-06-01").');
      expect(result).toBeDefined();
    });

    it('should handle state changes over time', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        happens(event(alice, became, intern), '2020-01-01').
        happens(event(alice, became, junior), '2021-01-01').
        happens(event(alice, became, senior), '2023-01-01').

        initiates(event(Person, became, NewRole), role(Person, NewRole), _).
        terminates(event(Person, became, NewRole), role(Person, OldRole), _) :-
          OldRole \\= NewRole.
      `);

      // Query for different time points
      const past = await prolog.queryOnce('holds_at(role(alice, intern), "2020-06-01").');
      const present = await prolog.queryOnce('holds_at(role(alice, senior), "2024-01-01").');

      expect(past).toBeDefined();
      expect(present).toBeDefined();
    });
  });

  describe('Complex temporal queries', () => {
    it('should find when events happened', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        happens(event(alice, learned, python), '2020-06-01').
        happens(event(alice, learned, javascript), '2021-03-01').
        happens(event(alice, learned, rust), '2023-09-01').
      `);

      const results = await collectAnswers(prolog.query('happens(event(alice, learned, Language), Date).'));
      expect(results).toBeDefined();
      expect(results.length).toBeGreaterThanOrEqual(3);
    });

    it('should compare event times', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        happens(event(alice, joined, company), '2020-01-01').
        happens(event(alice, became, manager), '2022-01-01').
        happens(event(alice, became, director), '2024-01-01').

        before(Event1, Event2) :-
          happens(Event1, Date1),
          happens(Event2, Date2),
          Date1 @< Date2.
      `);

      const result = await prolog.queryOnce('before(event(alice, joined, company), event(alice, became, manager)).');
      expect(result).toBeDefined();
      expect(result.status).toBe('success');
    });

    it('should handle concurrent states', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        happens(event(alice, started, project_a), '2024-01-01').
        happens(event(alice, started, project_b), '2024-02-01').
        happens(event(alice, finished, project_a), '2024-06-01').

        initiates(event(Person, started, Project), working_on(Person, Project), _).
        terminates(event(Person, finished, Project), working_on(Person, Project), _).
      `);

      // Check concurrent projects in March
      const results = await collectAnswers(prolog.query('holds_at(working_on(alice, Project), "2024-03-01").'));
      expect(results).toBeDefined();
      const successResults = results.filter((r): r is Extract<Answer, { status: 'success' }> => r.status === 'success');
      if (successResults.length > 0) {
        // Should be working on project_b at least
        const projects = successResults.map(r => {
          const proj = r.answer.Project;
          // Term could be an Atom object or a string
          return typeof proj === 'object' && proj !== null && 'functor' in proj ? proj.functor : String(proj);
        });
        expect(projects).toContain('project_b');
      }
    });
  });

  describe('Japanese language support', () => {
    it('should handle Japanese text in facts', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        happens(event('太郎', became, 'エンジニア'), '2024-01-01').
        happens(event('花子', learned, 'Python'), '2024-02-01').

        initiates(event(Person, became, Role), role(Person, Role), _).
      `);

      const results = await collectAnswers(prolog.query(`happens(event('太郎', became, Role), Date).`));
      expect(results).toBeDefined();
    });
  });

  describe('Rule learning integration', () => {
    it('should apply learned rules', async () => {
      if (!prolog) return;

      // Simulate learned rules from DynamicRuleLearner
      await prolog.consultText(`
        % Learned rule: 'learned' initiates 'knows'
        initiates(event(Person, learned, Subject), knows(Person, Subject), _).

        % Learned rule: 'married' initiates mutual relationship
        initiates(event(Person1, married, Person2), married_to(Person1, Person2), _).
        initiates(event(Person1, married, Person2), married_to(Person2, Person1), _).

        % Test data
        happens(event(alice, learned, python), '2020-01-01').
        happens(event(alice, married, bob), '2021-06-01').
      `);

      const knows = await prolog.queryOnce('holds_at(knows(alice, python), "2024-01-01").');
      const married = await prolog.queryOnce('holds_at(married_to(alice, bob), "2024-01-01").');

      expect(knows).toBeDefined();
      expect(married).toBeDefined();
    });
  });

  describe('Error handling', () => {
    it('should handle invalid queries gracefully', async () => {
      if (!prolog) return;

      try {
        const results = await collectAnswers(prolog.query('invalid_predicate(X, Y, Z).'));
        // Should return empty or error result, not throw
        expect(results).toBeDefined();
        expect(Array.isArray(results)).toBe(true);
      } catch (error) {
        // If it throws, that's also acceptable
        expect(error).toBeDefined();
      }
    });

    it('should handle malformed dates', async () => {
      if (!prolog) return;

      await prolog.consultText(`
        happens(event(alice, became, ceo), 'not-a-date').
      `);

      const result = await prolog.queryOnce('happens(event(alice, became, ceo), Date).');
      expect(result).toBeDefined();
      // Should still work even with non-standard date format
    });
  });
});

File: tests/prolog/event_calculus_test.pl
%% =============================================================================
%% Event Calculus Test Suite
%% =============================================================================
%% Native Prolog tests for the WhenM Event Calculus engine
%% Run with: swipl -g "run_tests." -t halt tests/prolog/event_calculus_test.pl
%% =============================================================================

:- use_module(library(plunit)).
:- consult('../../prolog/event_calculus.pl').

%% =============================================================================
%% Test Fixtures and Helpers
%% =============================================================================

%% Helper to set up a clean test environment
setup_test_env :-
    retractall(happens(_, _)),
    retractall(current_date(_)),
    assertz(current_date("2025-01-01")).

%% Helper to assert multiple events
assert_events([]).
assert_events([happens(Event, Time)|Rest]) :-
    assertz(happens(Event, Time)),
    assert_events(Rest).

%% =============================================================================
%% Core Event Calculus Tests
%% =============================================================================

:- begin_tests(event_calculus_core).

test(basic_holds_at, [setup(setup_test_env)]) :-
    % User learns Python on 2024-01-01
    assertz(happens(learned(user, python), "2024-01-01")),
    % Check it holds at various times
    assertion(holds_at(knows(user, python), "2024-01-01")),
    assertion(holds_at(knows(user, python), "2024-06-01")),
    assertion(holds_at(knows(user, python), "2025-01-01")).

test(fluent_termination, [setup(setup_test_env)]) :-
    % User joins club, then quits
    assertz(happens(joined(user, chess_club), "2024-01-01")),
    assertz(happens(quit(user, chess_club), "2024-06-01")),
    % Should hold before quit
    assertion(holds_at(member_of(user, chess_club), "2024-03-01")),
    % Should not hold after quit
    assertion(\+ holds_at(member_of(user, chess_club), "2024-07-01")).

test(clipped_predicate, [setup(setup_test_env)]) :-
    % Test the clipped/3 predicate directly
    assertz(happens(joined(user, club), "2024-01-01")),
    assertz(happens(quit(user, club), "2024-06-01")),
    % Fluent is clipped between initiation and after termination
    assertion(clipped("2024-01-01", member_of(user, club), "2024-07-01")),
    % Not clipped if we check before termination
    assertion(\+ clipped("2024-01-01", member_of(user, club), "2024-05-01")).

test(holds_now_predicate, [setup(setup_test_env)]) :-
    % Set current date
    retractall(current_date(_)),
    assertz(current_date("2024-12-01")),
    % Add event before current date
    assertz(happens(learned(user, prolog), "2024-06-01")),
    % Should hold "now"
    assertion(holds_now(knows(user, prolog))).

:- end_tests(event_calculus_core).

%% =============================================================================
%% Pattern-Based Rules Tests
%% =============================================================================

:- begin_tests(pattern_rules).

test(started_pattern, [setup(setup_test_env)]) :-
    % Test started_ prefix creates accumulating fluent
    assertz(happens(started_knows(user, java), "2024-01-01")),
    assertion(holds_at(knows(user, java), "2024-06-01")).

test(stopped_pattern, [setup(setup_test_env)]) :-
    % Test stopped_ prefix terminates fluent
    assertz(happens(started_member_of(user, gym), "2024-01-01")),
    assertz(happens(stopped_member_of(user, gym), "2024-06-01")),
    assertion(holds_at(member_of(user, gym), "2024-03-01")),
    assertion(\+ holds_at(member_of(user, gym), "2024-07-01")).

test(became_pattern_singular, [setup(setup_test_env)]) :-
    % Test became_ creates singular fluent (replaces previous)
    assertz(happens(became_role(user, developer), "2024-01-01")),
    assertz(happens(became_role(user, tech_lead), "2024-06-01")),
    % Old role should not hold after new one
    assertion(\+ holds_at(role(user, developer), "2024-07-01")),
    % New role should hold
    assertion(holds_at(role(user, tech_lead), "2024-07-01")).

test(pattern_fluent_helper, [setup(setup_test_env)]) :-
    % Test the pattern_fluent/3 helper
    assertion(pattern_fluent(started_, started_knows(user, rust), knows(user, rust))),
    assertion(pattern_fluent(became_, became_role(user, manager), role(user, manager))),
    assertion(\+ pattern_fluent(started_, not_matching(user, x), _)).

test(pattern_fluent_wildcard_helper, [setup(setup_test_env)]) :-
    % Test wildcard matching for singular fluents
    assertion(pattern_fluent_wildcard(became_, became_role(user, manager), role(user, _))),
    % Should preserve subject, wildcard the value
    pattern_fluent_wildcard(became_, became_role(john, manager), Fluent),
    Fluent = role(john, _).

:- end_tests(pattern_rules).

%% =============================================================================
%% Semantic Event Rules Tests
%% =============================================================================

:- begin_tests(semantic_events).

test(knowledge_events, [setup(setup_test_env)]) :-
    assert_events([
        happens(learned(user, python), "2024-01-01"),
        happens(studied(user, javascript), "2024-02-01"),
        happens(mastered(user, rust), "2024-03-01"),
        happens(met(user, alice), "2024-04-01")
    ]),
    assertion(holds_at(knows(user, python), "2024-06-01")),
    assertion(holds_at(knows(user, javascript), "2024-06-01")),
    assertion(holds_at(knows(user, rust), "2024-06-01")),
    assertion(holds_at(knows(user, alice), "2024-06-01")).

test(membership_events, [setup(setup_test_env)]) :-
    assert_events([
        happens(joined(user, chess_club), "2024-01-01"),
        happens(left(user, chess_club), "2024-03-01"),
        happens(joined(user, book_club), "2024-02-01")
    ]),
    assertion(\+ holds_at(member_of(user, chess_club), "2024-06-01")),
    assertion(holds_at(member_of(user, book_club), "2024-06-01")).

test(possession_events, [setup(setup_test_env)]) :-
    assert_events([
        happens(bought(user, car), "2024-01-01"),
        happens(acquired(user, house), "2024-02-01"),
        happens(sold(user, car), "2024-06-01"),
        happens(got(user, bike), "2024-07-01")
    ]),
    assertion(\+ holds_at(has(user, car), "2024-08-01")),
    assertion(holds_at(has(user, house), "2024-08-01")),
    assertion(holds_at(has(user, bike), "2024-08-01")).

test(location_singular, [setup(setup_test_env)]) :-
    assert_events([
        happens(moved_to(user, tokyo), "2024-01-01"),
        happens(moved_to(user, osaka), "2024-06-01"),
        happens(relocated_to(user, kyoto), "2024-09-01")
    ]),
    % Only most recent location should hold
    assertion(\+ holds_at(lives_in(user, tokyo), "2024-10-01")),
    assertion(\+ holds_at(lives_in(user, osaka), "2024-10-01")),
    assertion(holds_at(lives_in(user, kyoto), "2024-10-01")).

test(relationship_events, [setup(setup_test_env)]) :-
    assert_events([
        happens(married(user, alice), "2024-01-01"),
        happens(divorced(user, alice), "2024-06-01"),
        happens(married(user, bob), "2024-09-01")
    ]),
    assertion(\+ holds_at(married_to(user, alice), "2024-07-01")),
    assertion(holds_at(married_to(user, bob), "2024-10-01")).

test(preference_events, [setup(setup_test_env)]) :-
    assert_events([
        happens(started_liking(user, coffee), "2024-01-01"),
        happens(started_disliking(user, tea), "2024-02-01"),
        happens(stopped_liking(user, coffee), "2024-06-01")
    ]),
    assertion(\+ holds_at(likes(user, coffee), "2024-07-01")),
    assertion(holds_at(dislikes(user, tea), "2024-07-01")).

test(pet_events_3arity, [setup(setup_test_env)]) :-
    assert_events([
        happens(got_pet(user, dog, rex), "2024-01-01"),
        happens(got_pet(user, cat, whiskers), "2024-02-01"),
        happens(lost_pet(user, dog, rex), "2024-06-01")
    ]),
    assertion(\+ holds_at(has_pet(user, dog, rex), "2024-07-01")),
    assertion(holds_at(has_pet(user, cat, whiskers), "2024-07-01")).

test(business_events, [setup(setup_test_env)]) :-
    assert_events([
        happens(started_business(user, startup_a), "2024-01-01"),
        happens(founded(user, startup_b), "2024-02-01"),
        happens(sold_business(user, startup_a), "2024-06-01"),
        happens(closed_business(user, startup_b), "2024-09-01")
    ]),
    assertion(\+ holds_at(owns(user, startup_a), "2024-10-01")),
    assertion(\+ holds_at(owns(user, startup_b), "2024-10-01")).

test(employment_singular, [setup(setup_test_env)]) :-
    assert_events([
        happens(hired_at(user, company_a), "2024-01-01"),
        happens(left_company(user, company_a), "2024-06-01"),
        happens(joined_company(user, company_b), "2024-07-01")
    ]),
    assertion(\+ holds_at(employed_at(user, company_a), "2024-08-01")),
    assertion(holds_at(employed_at(user, company_b), "2024-08-01")).

test(project_events, [setup(setup_test_env)]) :-
    assert_events([
        happens(started_project(user, project_a), "2024-01-01"),
        happens(started_project(user, project_b), "2024-02-01"),
        happens(finished_project(user, project_a), "2024-06-01"),
        happens(abandoned_project(user, project_b), "2024-07-01")
    ]),
    assertion(\+ holds_at(working_on(user, project_a), "2024-08-01")),
    assertion(\+ holds_at(working_on(user, project_b), "2024-08-01")).

test(learning_progression, [setup(setup_test_env)]) :-
    assert_events([
        happens(started_learning(user, spanish), "2024-01-01"),
        happens(finished_learning(user, spanish), "2024-06-01"),
        happens(started_learning(user, french), "2024-07-01")
    ]),
    % Spanish: not learning anymore, but knows it
    assertion(\+ holds_at(learning(user, spanish), "2024-08-01")),
    assertion(holds_at(knows(user, spanish), "2024-08-01")),
    % French: still learning
    assertion(holds_at(learning(user, french), "2024-08-01")),
    assertion(\+ holds_at(knows(user, french), "2024-08-01")).

:- end_tests(semantic_events).

%% =============================================================================
%% Query Helpers Tests
%% =============================================================================

:- begin_tests(query_helpers).

test(all_holding, [setup(setup_test_env)]) :-
    assert_events([
        happens(learned(user, python), "2024-01-01"),
        happens(joined(user, club), "2024-02-01"),
        happens(bought(user, car), "2024-03-01")
    ]),
    all_holding("2024-06-01", Fluents),
    assertion(member(knows(user, python), Fluents)),
    assertion(member(member_of(user, club), Fluents)),
    assertion(member(has(user, car), Fluents)).

test(all_events, [setup(setup_test_env)]) :-
    assert_events([
        happens(learned(user, python), "2024-01-01"),
        happens(joined(user, club), "2024-02-01")
    ]),
    all_events(Events),
    assertion(length(Events, 2)),
    assertion(member(happens(learned(user, python), "2024-01-01"), Events)).

test(events_between, [setup(setup_test_env)]) :-
    assert_events([
        happens(event1, "2024-01-01"),
        happens(event2, "2024-03-01"),
        happens(event3, "2024-06-01"),
        happens(event4, "2024-09-01")
    ]),
    events_between("2024-02-01", "2024-07-01", Events),
    assertion(length(Events, 2)),
    assertion(member(happens(event2, "2024-03-01"), Events)),
    assertion(member(happens(event3, "2024-06-01"), Events)).

test(ever_held, [setup(setup_test_env)]) :-
    assert_events([
        happens(learned(user, python), "2024-01-01"),
        happens(joined(user, club), "2024-02-01"),
        happens(quit(user, club), "2024-06-01")
    ]),
    assertion(ever_held(knows(user, python))),
    assertion(ever_held(member_of(user, club))),
    assertion(\+ ever_held(has(user, spaceship))).

test(fluent_timeline, [setup(setup_test_env)]) :-
    assert_events([
        happens(joined(user, club), "2024-01-01"),
        happens(quit(user, club), "2024-06-01"),
        happens(joined(user, club), "2024-09-01")
    ]),
    fluent_timeline(member_of(user, club), Timeline),
    assertion(member(event(started, "2024-01-01"), Timeline)),
    assertion(member(event(ended, "2024-06-01"), Timeline)),
    assertion(member(event(started, "2024-09-01"), Timeline)).

:- end_tests(query_helpers).

%% =============================================================================
%% Utility Predicates Tests
%% =============================================================================

:- begin_tests(utilities).

test(assert_retract_event, [setup(setup_test_env)]) :-
    assert_event(test_event, "2024-01-01"),
    assertion(happens(test_event, "2024-01-01")),
    retract_event(test_event, "2024-01-01"),
    assertion(\+ happens(test_event, "2024-01-01")).

test(clear_person, [setup(setup_test_env)]) :-
    assert_events([
        happens(learned(alice, python), "2024-01-01"),
        happens(joined(alice, club), "2024-02-01"),
        happens(learned(bob, java), "2024-03-01")
    ]),
    clear_person(alice),
    assertion(\+ happens(learned(alice, python), _)),
    assertion(\+ happens(joined(alice, club), _)),
    assertion(happens(learned(bob, java), "2024-03-01")).

test(clear_all, [setup(setup_test_env)]) :-
    assert_events([
        happens(event1, "2024-01-01"),
        happens(event2, "2024-02-01"),
        happens(event3, "2024-03-01")
    ]),
    clear_all,
    all_events(Events),
    assertion(Events = []).

test(date_comparison, [setup(setup_test_env)]) :-
    assertion(date_before("2024-01-01", "2024-06-01")),
    assertion(\+ date_before("2024-06-01", "2024-01-01")),
    assertion(date_after("2024-06-01", "2024-01-01")),
    assertion(date_between("2024-03-01", "2024-01-01", "2024-06-01")),
    assertion(\+ date_between("2024-09-01", "2024-01-01", "2024-06-01")).

:- end_tests(utilities).

%% =============================================================================
%% Edge Cases and Complex Scenarios
%% =============================================================================

:- begin_tests(edge_cases).

test(multiple_initiations_same_fluent, [setup(setup_test_env)]) :-
    % Multiple ways to gain the same knowledge
    assert_events([
        happens(learned(user, python), "2024-01-01"),
        happens(studied(user, python), "2024-02-01"),
        happens(mastered(user, python), "2024-03-01")
    ]),
    % Should still hold (accumulating fluent)
    assertion(holds_at(knows(user, python), "2024-06-01")).

test(termination_before_initiation, [setup(setup_test_env)]) :-
    % Terminating event before initiating event (should have no effect)
    assert_events([
        happens(quit(user, club), "2024-01-01"),
        happens(joined(user, club), "2024-06-01")
    ]),
    assertion(\+ holds_at(member_of(user, club), "2024-03-01")),
    assertion(holds_at(member_of(user, club), "2024-07-01")).

test(simultaneous_events, [setup(setup_test_env)]) :-
    % Events happening at the exact same time
    assert_events([
        happens(joined(user, club), "2024-01-01"),
        happens(quit(user, club), "2024-01-01")
    ]),
    % Termination should take precedence when both happen at same time
    assertion(\+ holds_at(member_of(user, club), "2024-01-01")).

test(complex_timeline, [setup(setup_test_env)]) :-
    % Complex user journey
    assert_events([
        happens(moved_to(user, tokyo), "2024-01-01"),
        happens(hired_at(user, company_a), "2024-01-15"),
        happens(started_learning(user, japanese), "2024-02-01"),
        happens(married(user, alice), "2024-03-01"),
        happens(bought(user, house), "2024-04-01"),
        happens(finished_learning(user, japanese), "2024-06-01"),
        happens(left_company(user, company_a), "2024-07-01"),
        happens(joined_company(user, company_b), "2024-07-15"),
        happens(moved_to(user, osaka), "2024-08-01")
    ]),
    % Check state at end of timeline
    assertion(holds_at(lives_in(user, osaka), "2024-12-01")),
    assertion(\+ holds_at(lives_in(user, tokyo), "2024-12-01")),
    assertion(holds_at(employed_at(user, company_b), "2024-12-01")),
    assertion(holds_at(knows(user, japanese), "2024-12-01")),
    assertion(\+ holds_at(learning(user, japanese), "2024-12-01")),
    assertion(holds_at(married_to(user, alice), "2024-12-01")),
    assertion(holds_at(has(user, house), "2024-12-01")).

test(empty_database, [setup(setup_test_env)]) :-
    clear_all,
    % No events means no fluents hold
    assertion(\+ holds_at(knows(user, anything), "2024-06-01")),
    all_holding("2024-06-01", Fluents),
    assertion(Fluents = []).

:- end_tests(edge_cases).

%% =============================================================================
%% Performance Tests (optional, can be skipped for quick runs)
%% =============================================================================

:- begin_tests(performance, [condition(true)]).

test(large_event_set, [setup(setup_test_env), timeout(10)]) :-
    % Generate 100 events
    forall(
        between(1, 100, N),
        (
            atom_concat(skill_, N, Skill),
            format(atom(Date), "2024-01-~|~`0t~d~2|", [N mod 28 + 1]),
            assertz(happens(learned(user, Skill), Date))
        )
    ),
    % Query should complete within reasonable time
    all_holding("2024-06-01", Fluents),
    assertion(length(Fluents, 100)).

test(complex_termination_chain, [setup(setup_test_env), timeout(10)]) :-
    % Create chain of join/quit events
    forall(
        between(1, 50, N),
        (
            atom_concat(club_, N, Club),
            N2 is N * 2,
            format(atom(JoinDate), "2024-01-~|~`0t~d~2|", [N mod 28 + 1]),
            format(atom(QuitDate), "2024-02-~|~`0t~d~2|", [N mod 28 + 1]),
            assertz(happens(joined(user, Club), JoinDate)),
            assertz(happens(quit(user, Club), QuitDate))
        )
    ),
    % All should be terminated
    all_holding("2024-06-01", Fluents),
    assertion(Fluents = []).

:- end_tests(performance).

%% =============================================================================
%% Test Runner
%% =============================================================================

run_tests :-
    run_tests(event_calculus_core),
    run_tests(pattern_rules),
    run_tests(semantic_events),
    run_tests(query_helpers),
    run_tests(utilities),
    run_tests(edge_cases),
    run_tests(performance).